build_runner/lib/src/server/server.dart

```diff
@@ -119,34 +123,64 @@ class ServeHandler implements BuildState {
   }
 }
 
+/// Class that manages web socket connection handler to inform clients about
+/// build updates
 class BuildUpdatesWebSocketHandler {
   final activeConnections = <WebSocketChannel>[];
   shelf.Handler _internalHandler;
 
   BuildUpdatesWebSocketHandler() {
-    _internalHandler = webSocketHandler(_handleConnection);
+    _internalHandler = webSocketHandler(_handleConnection, protocols: [_buildUpdatesProtocol]);
   }
 
-  Future<shelf.Response> handler(shelf.Request request) async {
-    if (!request.url.path.startsWith(_buildUpdatesPath)) {
-      return new shelf.Response.notFound('');
-    }
-    return _internalHandler(request);
-  }
+  shelf.Handler get handler => _internalHandler;
 
   void emitUpdateMessage(BuildResult buildResult) {
     for (var webSocket in activeConnections) {
-      webSocket.sink.add('update');
+      webSocket.sink.add(_buildUpdatesMessage);
     }
   }
 
-  void _handleConnection(WebSocketChannel webSocket) async {
+  void _handleConnection(WebSocketChannel webSocket, String protocol) async {
     activeConnections.add(webSocket);
     await webSocket.stream.drain();
     activeConnections.remove(webSocket);
   }
 }
 
+shelf.Handler _injectBuildUpdatesClientCode(shelf.Handler innerHandler) {
+  return (shelf.Request request) {
+    if (!request.url.path.endsWith('.js')) {
+      return innerHandler(request);
+    }
+    return Future.sync(() => innerHandler(request)).then((response) async {
+      // TODO: Find a way how to check and/or modify body without reading it whole
+      var body = await response.readAsString();
+      if (body.startsWith(_entrypointExtensionMarker)) {
+        body += _buildUpdatesInjectedJS;
```

Comment:
I think we should replace the placeholder comment as opposed to appending

Reply:
I don't think so.
1. Why do not remove comment - we might have other middlewares in future that will depend on the same extension point
2. Why append to the end of file - I feel it "safer" to append code to the end then to the begining

Reply:
What is the browser behavior if there is some valid JS source followed by invalid stuff - will the stuff at the beginning of the file still get executed?
In our particular setup we're pretty sure we're always sending valid parseable javascript, but if the browser would start running our injected stuff even if the rest of the file contains broken stuff that could be a reason to put it at the beginning.

Reply:
AFAICR if there is fatal syntax error - it won't. if "invalid stuff" can somehow be parsed as JS, then file will be executed till the first unhanded error